File: backend/model/backwork.go
package model

import (
	"backend/database"

	"gorm.io/gorm"
)

type Backwork struct {
	gorm.Model
	UUID           string `gorm:"type:varchar(255);not null;index" json:"uuid"`
	FileName       string `gorm:"type:varchar(255);not null;index" json:"fileName"`
	FileURL        string `gorm:"type:varchar(255);not null;" json:"url"`
	CourseID       uint   `gorm:"not null;index" json:"courseId"`
	ProfessorID    uint   `gorm:"not null;index" json:"professorId"`
	Verified       bool   `gorm:"type:boolean;default:false;index" json:"verified,omitempty"`
	UserID         uint   `gorm:"not null;index" json:"userId"`
	RelevanceScore int    `gorm:"type:int;default:0" json:"relevanceScore,omitempty"`
}

// save the backwork to the database
func (bw *Backwork) Save() (*Backwork, error) {
	err := database.Database.Create(&bw).Error

	if err != nil {
		return &Backwork{}, err
	}

	return bw, nil
}

File: frontend/src/Backwork.js
import React, { useEffect, useState } from 'react';
import axios from 'axios';

const BackworkList = () => {
  const [backworks, setBackworks] = useState([]);
  const jwt = localStorage.getItem('jwtToken');
  const headers = {
    Authorization: `Bearer ${jwt}`,
    'Content-Type': 'application/json',
  };

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await axios({
          method: 'post',
          url: 'http://localhost:8000/api/backworks',
          headers,
          data: {
            courseid: 1,
            professorid: null,
            verified: false,
          },
        });

        setBackworks(response.data.backworks);
      } catch (err) {
        console.error(err);
      }
    };

    fetchData();
  }, []);

  const handleDownload = async (url, filename) => {
    try {
      const response = await axios.get(url, { responseType: 'blob' });
      const blob = new Blob([response.data], { type: 'application/octet-stream' });
      const link = document.createElement('a');
      link.href = window.URL.createObjectURL(blob);
      link.download = filename;
      link.click();
    } catch (err) {
      console.error(err);
    }
  };

  return (
    <ul>
      {backworks.map((backwork) => (
        <li key={backwork.uuid}>
          <h2>{backwork.fileName}</h2>
          <button onClick={() => handleDownload(backwork.url, backwork.fileName)}>
            Download file
          </button>
          <p>Course ID: {backwork.courseId}</p>
          <p>Professor ID: {backwork.professorId}</p>
          <p>Verified: {backwork.verified ? 'Yes' : 'No'}</p>
          <p>User ID: {backwork.userId}</p>
        </li>
      ))}
    </ul>
  );
};

export default BackworkList;